Cluster mempool followups

sdaftuar opened the pull request:
As suggested in the main cluster mempool PR (https://github.com/bitcoin/bitcoin/pull/28676#pullrequestreview-3177119367), I've pulled out some of the non-essential optimizations and cleanups into this separate PR.  

Will continue to add more commits here to address non-blocking suggestions/improvements as they come up.

sdaftuar committed:
Remove unused argument to RemoveStaged

sdaftuar committed:
Remove unused variable (cacheMap) in mempool

sdaftuar committed:
Invoke removeUnchecked() directly in removeForBlock()

sdaftuar committed:
scripted-diff: rename AddToMempool -> TryAddToMempool

-BEGIN VERIFY SCRIPT-
find src/test -type f -exec sed -i 's/AddToMempool/TryAddToMempool/g' {} +
find src/bench -type f -exec sed -i 's/AddToMempool/TryAddToMempool/g' {} +
-END VERIFY SCRIPT-

sdaftuar committed:
Simplify removeRecursive

sdaftuar committed:
Rewrite removeForReorg to avoid using sets

Also improve test coverage for removeForReorg by creating a scenario where
there are in-mempool descendants that are only invalidated due to an in-mempool
parent no longer spending a mature coin.

sdaftuar committed:
Rewrite GetChildren without sets

sdaftuar committed:
Warn user if using -limitancestorsize/-limitdescendantsize that the options have no effect

Also update the help text for -limitancestorcount/-limitdescendantcount to
explain they no longer affect the mempool, and are only used by the wallet for
coin selection.

sdaftuar committed:
Remove ancestor and descendant vsize limits from MemPoolLimits

sdaftuar committed:
Use cluster limits instead of ancestor/descendant limits when sanity checking package policy limits

sdaftuar committed:
Use cluster size limit instead of ancestor/descendant size limits when sanity checking TRUC policy limits

sdaftuar committed:
Use cluster size limit instead of ancestor size limit in txpackage unit test

sdaftuar committed:
Remove unused DEFAULT_ANCESTOR_SIZE_LIMIT_KVB and DEFAULT_DESCENDANT_SIZE_LIMIT_KVB

sdaftuar committed:
Add a GetFeePerVSize() accessor to CFeeRate, and use it in the BlockAssembler

sdaftuar committed:
miner: replace "package" with "chunk"

This makes the terminology consistent with other parts of the codebase, as part
of the cluster mempool implementation.

sdaftuar committed:
doc: Add design notes for cluster mempool and explain new mempool limits

sdaftuar committed:
doc: add release notes snippet for cluster mempool

sdaftuar committed:
Avoid using mapTx.modify() to update modified fees

Now that the mempool no longer keeps any feerate-based indices, we can modify
feerates in mempool entries directly.

sdaftuar committed:
rpc: improve getmempoolcluster output

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -261,14 +261,14 @@ static RPCHelpMan testmempoolaccept()
 {
     return {
-        RPCResult{RPCResult::Type::NUM, "vsize", "virtual transaction size as defined in BIP 141. This is different from actual serialized size for witness transactions as witness data is discounted."},
+        RPCResult{RPCResult::Type::NUM, "weight", "total sigops-adjusted weight (as defined in BIP 141 and modified by '-bytespersigop'"},
```
nit: missing parens

Also, I wonder if it would be confusing that an RPC result named "weight" is adjusted in some RPCs but not in others... should we be calling this "adjusted_weight" ? Or perhaps "clusterweight" and "chunkweight" are enough to indicate that this is in the mempool context and thus not just BIP 141 weight.

glozow replied:
Some documentation that could be worth having in doc/policy/ ?

### Fee and Size Terminology in Mempool Policy

- Each transaction has a **weight** and virtual size as defined in BIP 141 (different from serialized size for witness transactions as witness data is discounted and the value is rounded up to the nearest integer).

    - In the RPCs, "weight" refers to the weight as defined in BIP 141. (Not really true here since chunk weight is sigops-adjusted)

- A transaction has a **sigops size**, defined as its sigop cost multiplied by the node's `-bytespersigop`, an adjustable policy.

- A transaction's **virtual size (vsize)** refers to its **sigops-adjusted virtual size**: the maximum of its BIP 141 size and sigop size. This virtual size is used in order to simplify the process of building blocks that satisfy both the maximum weight limit and sigop limit.

    - In the RPCs, "vsize" refers to sigops-adjusted virtual size.

    - Mempool entry information with the suffix "size" ("clustersize") refer to the cummulative size of the transactions in the group.

- A transaction can also have a **sigops-adjusted weight**, defined similarly (using the BIP 141 weight without dividing by the witness scaling factor). This value is used in most internal processes because using virtual size can result in loss of precision.

- A transaction's **base fee** is the difference between its input and output values.

- A transaction's **modified fee** is its base fee added to any **fee delta** introduced using the `prioritisetransaction` RPC.
Modified fee is used internally for all fee-related mempool policies and block building.

sdaftuar replied:
I'm going with `chunkweight` and `clusterweight` for now. Not sure what to do for the output of `getmempoolfeeratediagram()` though?

I also took your doc suggestion and created a new file in doc/policy/ that explains the terminology. Would it better to consolidate some of those files into one?

instagibbs commented on src/txmempool.cpp:
```diff
@@ -308,35 +308,49 @@ CTxMemPool::txiter CTxMemPool::CalculateDescendants(const CTxMemPoolEntry& entry
+        removeRecursive(origit, reason);
+    } else {
+        // When recursively removing but origTx isn't in the mempool
+        // be sure to remove any children that are in the pool. This can
```
f1cfc07b8af90d9c80c84e9db56d61c56f50d6c2

```Suggestion
        // be sure to remove any descendants that are in the pool. This can
```

instagibbs commented on src/txmempool.cpp:
```diff
@@ -308,35 +308,49 @@ CTxMemPool::txiter CTxMemPool::CalculateDescendants(const CTxMemPoolEntry& entry
-                txToRemove.insert(nextit);
+    txiter origit = mapTx.find(origTx.GetHash());
+    if (origit != mapTx.end()) {
+        removeRecursive(origit, reason);
```
f1cfc07b8af90d9c80c84e9db56d61c56f50d6c2

From a cursory glance I could not figure out when this happens. I believe behavior matches, but I'm still curious.

sdaftuar replied:
In `removeConflicts()`, we'd invoke `removeRecursive()` with a transaction that is in the mempool, and then this code would trigger.  However I realize now that we already have the iterator that we're trying to remove, so I can save a map lookup by invoking the helper directly.  (Prior to #33629 there were more places we'd invoke `removeRecursive()` on transactions in the mempool.)

Otherwise, I think this code will now only trigger in tests.

instagibbs commented on src/txmempool.cpp:
```diff
@@ -360,15 +360,25 @@ void CTxMemPool::removeForReorg(CChain& chain, std::function<bool(txiter)> check
+    {
+        WITH_FRESH_EPOCH(m_epoch);
+        for (indexed_transaction_set::const_iterator it = mapTx.begin(); it != mapTx.end(); it++) {
+            if (check_final_and_mature(it) && !visited(it)) {
```
05a7b963c93501939ec0523fe0182af87ee40056

Light preference for splitting out the checks to make the `visited()` mutation more obvious

```
diff --git a/src/txmempool.cpp b/src/txmempool.cpp
index 5efdfbf133..7765f27b79 100644
--- a/src/txmempool.cpp
+++ b/src/txmempool.cpp
@@ -365,5 +365,6 @@ void CTxMemPool::removeForReorg(CChain& chain, std::function<bool(txiter)> check
         WITH_FRESH_EPOCH(m_epoch);
         for (indexed_transaction_set::const_iterator it = mapTx.begin(); it != mapTx.end(); it++) {
-            if (check_final_and_mature(it) && !visited(it)) {
+            if (check_final_and_mature(it)) {
+                if (visited(it)) continue;
                 txToRemove.emplace_back(it);
                 auto descendants = m_txgraph->GetDescendants(*it, TxGraph::Level::MAIN);
```

sdaftuar committed:
Sanity check feerate diagram in CTxMemPool::check()

Also switch Assume()'s to assert()'s, so that failures in this function are
always caught.

sdaftuar committed:
Fix comment to reference cluster limits, not chain limits

sdaftuar committed:
Improve comments for getTransactionAncestry to reference cluster counts instead of descendants

sdaftuar committed:
Require mempool lock to be held when invoking TRUC checks

instagibbs commented on src/node/miner.cpp:
```diff
@@ -252,7 +252,7 @@ void BlockAssembler::addChunks()
     while (selected_transactions.size() > 0) {
         // Check to see if min fee rate is still respected.
-        if (chunk_feerate.fee < m_options.blockMinFeeRate.GetFee(chunk_feerate_vsize.size)) {
+        if (chunk_feerate_vsize < m_options.blockMinFeeRate.GetFeePerVSize()) {
```
ff2a16af040ac0906b4ebd488b1125b6ea25fa14

I don't think we should be tie-breaking based on size here?

```Suggestion
        if (chunk_feerate_vsize << m_options.blockMinFeeRate.GetFeePerVSize()) {
```

sdaftuar committed:
Rename weight -> clusterweight in RPC output, and add doc explaining mempool terminology

Co-authored-by: glozow <gloriajzhao@gmail.com>

sdaftuar committed:
Fix parameter name in getmempoolcluster rpc

sdaftuar committed:
doc: add comment to explain correctness of GatherClusters()

instagibbs commented on src/rpc/mempool.cpp:
```diff
@@ -320,16 +334,28 @@ static void clusterToJSON(const CTxMemPool& pool, UniValue& info, std::vector<co
-        txentry.pushKV("chunkfee", ValueFromAmount((int)feerate.fee));
-        txentry.pushKV("chunkweight", feerate.size);
-        txs.push_back(txentry);
+        if (current_chunk_feerate.size == 0) {
```
486b441971912bdc63921425d5e40a8c168186ea

pretty sure there's at least one bug here. f.e., if you have three chunks in one cluster, I think the conditional here is only hit the first time, as `current_chunk_feerate` is never reset to the next chunk's size, so it'll just start going negative. 

Wrote a test which I think covers it

```
diff --git a/test/functional/mempool_cluster.py b/test/functional/mempool_cluster.py
index 4b4812e619..a7eabef039 100755
--- a/test/functional/mempool_cluster.py
+++ b/test/functional/mempool_cluster.py
@@ -299,4 +299,73 @@ class MempoolClusterTest(BitcoinTestFramework):
         assert_equal(node.getmempoolcluster(tx_replacer["txid"])['txcount'], 2)
 
+    @cleanup
+    def test_getmempoolcluster(self):
+        node = self.nodes[0]
+
+        self.log.info("Testing getmempoolcluster")
+
+        assert_equal(node.getrawmempool(), [])
+
+        # Not in-mempool
+        not_mempool_tx = self.wallet.create_self_transfer()
+        assert_raises_rpc_error(-5, "Transaction not in mempool", node.getmempoolcluster, not_mempool_tx["txid"])
+
+        # Test that chunks are being recomputed properly
+
+        # One chunk with one tx
+        first_chunk_tx = self.wallet.send_self_transfer(from_node=node)
+        first_chunk_info = node.getmempoolcluster(first_chunk_tx["txid"])
+        assert_equal(first_chunk_info, {'weight': first_chunk_tx["tx"].get_weight(), 'txcount': 1, 'chunks': [{'chunkfee': first_chunk_tx["fee"], 'chunkweight': first_chunk_tx["tx"].get_weight(), 'txs': [first_chunk_tx["txid"]]}]})
+
+        # Another unconnected tx, nothing should change
+        other_chunk_tx = self.wallet.send_self_transfer(from_node=node)
+        first_chunk_info = node.getmempoolcluster(first_chunk_tx["txid"])
+        assert_equal(first_chunk_info, {'weight': first_chunk_tx["tx"].get_weight(), 'txcount': 1, 'chunks': [{'chunkfee': first_chunk_tx["fee"], 'chunkweight': first_chunk_tx["tx"].get_weight(), 'txs': [first_chunk_tx["txid"]]}]})
+
+        # Second connected tx, makes one chunk still with high enough fee
+        second_chunk_tx = self.wallet.send_self_transfer(from_node=node, utxo_to_spend=first_chunk_tx["new_utxo"], fee_rate=Decimal("0.01"))
+        first_chunk_info = node.getmempoolcluster(first_chunk_tx["txid"])
+        # output is same across same cluster transactions
+        assert_equal(first_chunk_info, node.getmempoolcluster(second_chunk_tx["txid"]))
+        chunkweight = first_chunk_tx["tx"].get_weight() + second_chunk_tx["tx"].get_weight()
+        chunkfee = first_chunk_tx["fee"] + second_chunk_tx["fee"]
+        assert_equal(first_chunk_info, {'weight': chunkweight, 'txcount': 2, 'chunks': [{'chunkfee': chunkfee, 'chunkweight': chunkweight, 'txs': [first_chunk_tx["txid"], second_chunk_tx["txid"]]}]})
+
+        # Third connected tx, makes one chunk still with high enough fee
+        third_chunk_tx = self.wallet.send_self_transfer(from_node=node, utxo_to_spend=second_chunk_tx["new_utxo"], fee_rate=Decimal("0.1"))
+        first_chunk_info = node.getmempoolcluster(first_chunk_tx["txid"])
+        # output is same across same cluster transactions
+        assert_equal(first_chunk_info, node.getmempoolcluster(third_chunk_tx["txid"]))
+        chunkweight = first_chunk_tx["tx"].get_weight() + second_chunk_tx["tx"].get_weight() + third_chunk_tx["tx"].get_weight()
+        chunkfee = first_chunk_tx["fee"] + second_chunk_tx["fee"] + third_chunk_tx["fee"]
+        assert_equal(first_chunk_info, {'weight': chunkweight, 'txcount': 3, 'chunks': [{'chunkfee': chunkfee, 'chunkweight': chunkweight, 'txs': [first_chunk_tx["txid"], second_chunk_tx["txid"], third_chunk_tx["txid"]]}]})
+
+        # Now test single cluster with each tx being its own chunk
+
+        # One chunk with one tx
+        first_chunk_tx = self.wallet.send_self_transfer(from_node=node)
+        first_chunk_info = node.getmempoolcluster(first_chunk_tx["txid"])
+        assert_equal(first_chunk_info, {'weight': first_chunk_tx["tx"].get_weight(), 'txcount': 1, 'chunks': [{'chunkfee': first_chunk_tx["fee"], 'chunkweight': first_chunk_tx["tx"].get_weight(), 'txs': [first_chunk_tx["txid"]]}]})
+
+        # Second connected tx, lower fee
+        second_chunk_tx = self.wallet.send_self_transfer(from_node=node, utxo_to_spend=first_chunk_tx["new_utxo"], fee_rate=Decimal("0.000002"))
+        first_chunk_info = node.getmempoolcluster(first_chunk_tx["txid"])
+        # output is same across same cluster transactions
+        assert_equal(first_chunk_info, node.getmempoolcluster(second_chunk_tx["txid"]))
+        first_chunkweight = first_chunk_tx["tx"].get_weight()
+        second_chunkweight = second_chunk_tx["tx"].get_weight()
+        assert_equal(first_chunk_info, {'weight': first_chunkweight + second_chunkweight, 'txcount': 2, 'chunks': [{'chunkfee': first_chunk_tx["fee"], 'chunkweight': first_chunkweight, 'txs': [first_chunk_tx["txid"]]}, {'chunkfee': second_chunk_tx["fee"], 'chunkweight': second_chunkweight, 'txs': [second_chunk_tx["txid"]]}]})
+
+        # Third connected tx, even lower fee
+        third_chunk_tx = self.wallet.send_self_transfer(from_node=node, utxo_to_spend=second_chunk_tx["new_utxo"], fee_rate=Decimal("0.000001"))
+        first_chunk_info = node.getmempoolcluster(first_chunk_tx["txid"])
+        # output is same across same cluster transactions
+        assert_equal(first_chunk_info, node.getmempoolcluster(third_chunk_tx["txid"]))
+        first_chunkweight = first_chunk_tx["tx"].get_weight()
+        second_chunkweight = second_chunk_tx["tx"].get_weight()
+        third_chunkweight = third_chunk_tx["tx"].get_weight()
+        chunkfee = first_chunk_tx["fee"] + second_chunk_tx["fee"] + third_chunk_tx["fee"]
+        # FIXME I think there's a big consolidating the second and third tx into one chunk
+        assert_equal(first_chunk_info, {'weight': first_chunkweight + second_chunkweight + third_chunkweight, 'txcount': 3, 'chunks': [{'chunkfee': first_chunk_tx["fee"], 'chunkweight': first_chunkweight, 'txs': [first_chunk_tx["txid"]]}, {'chunkfee': second_chunk_tx["fee"], 'chunkweight': second_chunkweight, 'txs': [second_chunk_tx["txid"]]}, {'chunkfee': third_chunk_tx["fee"], 'chunkweight': third_chunkweight, 'txs': [third_chunk_tx["txid"]]}]})
 
     def run_test(self):
@@ -305,4 +374,6 @@ class MempoolClusterTest(BitcoinTestFramework):
         self.generate(self.wallet, 400)
 
+        self.test_getmempoolcluster()
+
         self.test_cluster_limit_rbf(DEFAULT_CLUSTER_LIMIT)
 ```

instagibbs reviewed (commented):
reviewed through 875f1b98fcc9a992a3280bf32edc2d30c1fb034d

sipa commented on src/txmempool.cpp:
```diff
@@ -308,35 +308,49 @@ CTxMemPool::txiter CTxMemPool::CalculateDescendants(const CTxMemPoolEntry& entry
+        // the mempool for any reason.
+        auto iter = mapNextTx.lower_bound(COutPoint(origTx.GetHash(), 0));
+        std::vector<CTxMemPool::txiter> to_remove;
+        WITH_FRESH_EPOCH(m_epoch);
```
In commit "Simplify removeRecursive"

I don't think it matters for performance, but I think it's simpler to:
* first build a `std::vector<TxGraph::Ref*>` for all mempool entries corresponding to direct in-mempool children
* then pass that to `TxGraph::GetDescendantsUnion`
* then invoke `removeUnchecked` for each of the results

sipa commented on doc/policy/mempool-limits.md:
```diff
@@ -8,58 +8,12 @@ Tx0 is a *parent* of Tx1 and Tx1 is a *child* of Tx0.
-A mempool entry's *descendant count* is the total number of in-mempool (unconfirmed) transactions in
-its descendant set, including itself.
+A transaction's *cluster* is the set of all transactions that can be reached
+from that transaction by following all parent or child relationships. For
```
In commit "doc: Update mempool-limits.md for cluster mempool".

Perhaps worth elaborating that it is not just parents and children, but also any combination thereof:

> A transaction's *cluster* is the set of all transactions that can be reached by following any combination of parent or child relationships. This includes all ancestors and descendants, but also any descendants of those ancestors and ancestors of those descendants, and so on.

sipa commented on doc/policy/mempool-design.md:
```diff
@@ -0,0 +1,19 @@
+
+## Definitions and theorems
+
+See https://delvingbitcoin.org/t/cluster-mempool-definitions-theory/202 for
```
In commit "doc: add design notes for cluster mempool"

I think the theory thread is outdated now. Merging and postlinearization won't be very relevant post SFL, and the existence of optimal linearizations can probably be shown much more concisely when using results from the GGT paper.

It's probably worth creating a trimmed-down version with just definitions of clusters, chunks, and fee-size diagams, possibly to include directly in the document here, but that can be done as a follow-up I think.

sdaftuar replied:
Made an attempt at this in 9d25e2cbcd011b3046f9c3cd8de4e1b9112d0dfd

sdaftuar commented on src/txmempool.cpp:
```diff
@@ -360,15 +360,25 @@ void CTxMemPool::removeForReorg(CChain& chain, std::function<bool(txiter)> check
     }
-    RemoveStaged(setAllRemoves, MemPoolRemovalReason::REORG);
     for (indexed_transaction_set::const_iterator it = mapTx.begin(); it != mapTx.end(); it++) {
         assert(TestLockPointValidity(chain, it->GetLockPoints()));
```
I agree that this seems expensive, but an `Assume()` wouldn't save us anything, would it? I believe the code would still get invoked (unless the compiler is somehow smart enough to optimize it out).

instagibbs replied:
Given the new implementation that doesn't even use `visited()`, I think can be removed anyway (in another followup)

sdaftuar commented on src/txmempool.h:
```diff
@@ -320,6 +320,7 @@ class CTxMemPool
      */
     void check(const CCoinsViewCache& active_coins_tip, int64_t spendheight) const EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
 
+    void removeRecursive(txiter to_remove, MemPoolRemovalReason reason) EXCLUSIVE_LOCKS_REQUIRED(cs);
```
Made private in 6797c71ef9323efd2db9dc07d6c1156d077b38cc

sdaftuar commented on doc/policy/mempool-limits.md:
```diff
@@ -8,58 +8,12 @@ Tx0 is a *parent* of Tx1 and Tx1 is a *child* of Tx0.
-
-2. The descendant size limit temporarily is increased by the virtual size of the to-be-replaced
-   directly conflicting transaction.
+Transactions submitted to the mempool must not exceed the cluster limits set by
```
I ended up reworking the docs and merged the mempool-limits.md file (which was quite short and included a few definitions) in with the new mempool-design.md file (that also has a bunch of definitions and a brief overview of how cluster mempool works).

I took your language suggestion and incorporated in 9d25e2cbcd011b3046f9c3cd8de4e1b9112d0dfd.

sipa commented on src/txmempool.cpp:
```diff
@@ -352,15 +352,21 @@ void CTxMemPool::removeForReorg(CChain& chain, std::function<bool(txiter)> check
-    for (indexed_transaction_set::const_iterator it = mapTx.begin(); it != mapTx.end(); it++) {
-        if (check_final_and_mature(it)) txToRemove.insert(it);
+    std::vector<const TxGraph::Ref*> txToRemove;
+    {
```
In commit "Rewrite removeForReorg to avoid using sets"

Nit: it seems this scope level is unnecessary.

sipa commented on doc/policy/mempool-design.md:
```diff
@@ -0,0 +1,105 @@
+
+We view the unconfirmed transactions in the mempool as a directed graph,
+with an edge from transaction B to transaction A if B spends an output created
+by A (ie, B is a **child** of A, and A is a **parent** of B).
```
In commit "doc: Add design notes for cluster mempool and explain new mempool limits"

Writing nit: `ie,` -> `i.e.,`

sipa commented on doc/release-notes-33629.md:
```diff
@@ -0,0 +1,41 @@
+
+- Within the mempool, transactions are ordered based on the fee rate at which
+  they are expected to be mined, which takes into account the full set, or
+  "chunk", of transactions that would be included together (eg a parent and its
```
In commit "doc: add release notes snippet for cluster mempool"

Writing nit: `eg` -> `e.g.,`.

sdaftuar replied:
Fixed.

sipa commented on src/txmempool.cpp:
```diff
@@ -440,12 +442,27 @@ void CTxMemPool::check(const CCoinsViewCache& active_coins_tip, int64_t spendhei
     // Number of chunks is bounded by number of transactions.
     const auto diagram{GetFeerateDiagram()};
     Assume(diagram.size() <= score_with_topo.size() + 1);
+    Assume(diagram.size() >= 1);
```
In commit "Sanity check feerate diagram in CTxMemPool::check()"

In a check function like `CTxMempool::check()`, I think we should always use `assert`/`Assert`, not `Assume`, as we want the checks to always be enabled.

sdaftuar commented on doc/policy/mempool-design.md:
```diff
@@ -0,0 +1,105 @@
+### Motivation
+
+Selecting chunks in decreasing feerate order when building a block template
+will be close to optimal when the maximize size of a chunk is small compared to
```
Yep, that is what I meant!  Fixed.

sipa reviewed (commented):
ACK 48c29fbe0974664e28564af07fbfd753cf125c36

See nits above.

instagibbs reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/33591/commits/b8d279a81c16fe9f5b6d422e518c77344e217d4f

sipa reviewed (commented):
ACK b8d279a81c16fe9f5b6d422e518c77344e217d4f

glozow commented on src/policy/packages.h:
```diff
@@ -24,15 +24,10 @@ static constexpr uint32_t MAX_PACKAGE_COUNT{25};
-static_assert(DEFAULT_ANCESTOR_LIMIT >= MAX_PACKAGE_COUNT);
-static_assert(MAX_PACKAGE_WEIGHT >= DEFAULT_ANCESTOR_SIZE_LIMIT_KVB * WITNESS_SCALE_FACTOR * 1000);
-static_assert(MAX_PACKAGE_WEIGHT >= DEFAULT_DESCENDANT_SIZE_LIMIT_KVB * WITNESS_SCALE_FACTOR * 1000);
+// Packages are part of a single cluster, so ensure that the package limits are
```
Agree the previous checks can be removed. Rationale is "transactions that would be individually accepted may be rejected in a package erroneously" which is addressed more comprehensively by trying individual transactions if they don't need to be grouped with others and calculating precise cluster limits through the staging txgraph.
